### DISCUSSION TITLE
Is it possible to isolate layers from each other?

### DISCUSSION THREAD

**Discussion by ChrisGV04** [nuxt/nuxt#29676] [Questions]
Is it possible to create isolated layers such that I can split different areas of my app each with their own components, pages and utils while maintaining a common base?

## What I'm trying to do

I want to build an ecommerce website that also has a CMS as an admin area. Both are separate apps in the sense that they have their own UI styles and very different use-cases.

The CMS is an SPA an all of it's components are more "admin looking". The ecommerce site is the public website that uses SSR and follows the visual identity of the brand.

Both apps must also share certain utils from a base Nuxt App. For example: auth store, zod schemas, general purpose components, shared utils, etc. However, each Nuxt app has their own `UiButton.vue` component which looks and works differently per app.

I'm thinking the folder structure would look something like this:

```
app <- This is the "base" Nuxt app to share common assets between layers
--| components
----| BrandLogo.vue
--| composables
----| useAuth.ts

layers
--| cms <- This is the "admin" layer
----| app
------| components
--------| UiButton.vue <- This button has an admin style and specific props
------| pages
--------| admin
----------| index.vue
----| nuxt.config.ts
--| website <- This is the public website layer
----| app
------| components
--------| UiButton.vue <- This button has a branded style with reduced props
------| pages
--------| index.vue
----| nuxt.config.ts
```

Currently, if I use `<UiButton>` inside of my `website` layer it renders the one from the `cms` layer, which is not what I intended.

I also don't want to have auto-imports between layers, so that a composable inside the `cms` layer should not be available inside the `website` layer.

I don't know if this is at all possible with Nuxt layers or with a monorepo, so any help is greatly appreciated!

**Comment by harlio** [comment]
also interested in a similar type of project structure, did you ever figure out how to accomplish this?

**Comment by ChrisGV04** [comment]
Not exactly, but I achieved something I felt comfortable enough to use, which is using per-route-components and disabling component auto-imports. In my `nuxt.config.ts` I added the following:

```ts
export default defineNuxtConfig({
  components: false, // Disable component auto-imports

  pages: {
    pattern: [
      '**/*.vue',
      '!**/_components/**', // Ignore page-specific components
      '!**/_utils/**', // Ignore page-specific utilities
    ],
  },
})
```

That way I can have components and utilities next to my pages and they wont be registered as routes. For example:

```
--| pages
----| admin
------| _components <- Components that are specific for `admin` routes
--------| AdminSpecificComponent.vue
------| _utils <- Utilities that are specific for `admin` routes
--------| admin-composables.ts
------| index.vue <- Admin page that imports from `./_components/AdminSpecificComponent.vue`
----| website
------| _components <- Components that are specific for `website` routes
--------| WebsiteSpecificComponent.vue
------| index.vue <- Website page that imports from `./_components/WebsiteSpecificComponent.vue`
```

The only sacrifice is that I'm no longer using auto-imports, but I don't miss it since it is not as useful in very large projects.

**Comment by cernymatej** [comment]
When you use local layers, Nuxt assumes that they are a part of the same project. If you want to have isolated projects, I'd recommend using a monorepo and adding the base layer to the projects via `extends` in `nuxt.config.ts`.

**Comment by nowo** [comment]
This should suit you.[https://github.com/nuxt/example-layers-monorepo](https://github.com/nuxt/example-layers-monorepo)

**Comment by ojvribeiro** [comment]
I always wondered that, if I can have, say, an `utils` layer which I can use I other layers (like a _dev dependency layer_), that are used by the main project. Right now the terminal goes crazy saying the the util from layer `x` was ignored and the same util from layer `y` was used, multiple times. A monorepo does not really apply for my use case, as layers could be created by anyone to extend the main layer.
